tick/preprocessing/longitudinal_features_product.py

```diff
@@ -0,0 +1,236 @@
+
+        Parameters
+        ----------
+        X : `[np.ndarrays]`  or `[csr_matrices]`, shape=(n_intervals, n_features)
+            The list of features matrices.
+
+        Returns
+        -------
+        output : `LongitudinalFeaturesProduct`
+            The fitted current instance.
+        """
+        self.reset()
+        # TODO: check input type ?
+        # TODO: Check that all the matrices have the same size ?
+        n_intervals, n_init_features = X[0].shape
+        if n_init_features < 2:
+            raise ValueError("There should at least two features to compute\
+                     product features.")
+        self._set("_n_init_features", n_init_features)
+        self._set("_n_intervals", n_intervals)
+        comb_it = combinations(range(n_init_features), 2)
+        mapper = {i + n_init_features: c for i, c in enumerate(comb_it)}
+        self._set("_mapper", mapper)
+        self._set("_n_output_features", int(n_init_features +
+                                            comb(n_init_features, 2)))
+
+        if sps.issparse(X[0]) and self.exposure_type == "infinite":
+            self._set("_preprocessor", SparseLongitudinalFeaturesProduct(X))
+
+        self._set("_fitted", True)
+
+        return self
+
+    def transform(self, X):
+        """Add the product features to the given features matrices list.
+
+        Parameters
+        ----------
+        X : `[np.ndarrays]`  or `[csr_matrices]`, shape=(n_intervals, n_features)
+            The list of features matrices.
+
+        Returns
+        -------
+        output : `[np.ndarrays]`  or `[csr_matrices]`, shape=(n_intervals, n_features)
+            The list of features matrices with added product features.
+        """
+
+        # TODO: check matrix dimensions and consistency with self.fit ?
+        if self.exposure_type == "short":
+            X_with_products = self.short_exposure_products(X)
+        elif self.exposure_type == "infinite":
+            X_with_products = self.infinite_exposure_products(X)
+
+        return X_with_products
+
+    def fit_transform(self, X):
+        """Fit and apply the product feature computation using the features 
+        matrices list.
+
+        Parameters
+        ----------
+        X : `[np.ndarrays]`  or `[csr_matrices]`, shape=(n_intervals, n_features)
+            The list of features matrices.
+
+        Returns
+        -------
+        output : `pd.DataFrame`
+            The binarized features matrix. The number of columns is
+            larger than n_features, smaller than n_cuts * n_features,
+            depending on the actual number of columns that have been
+            binarized.
+        """
+        self.fit(X)
+        X_with_products = self.transform(X)
+
+        return X_with_products
+
+    def infinite_exposure_products(self, X):
+        if sps.issparse(X[0]):
+            X_with_products = [self._sparse_infinite_product(arr) for arr in X]  # TODO: fix multiprocessing
```

Comment:
**Remark**
This loop should be done in C++, but we can keep it in python for now.
Multiprocessing should and could be done from C++ as well... but later also :)
In general, maybe many things done here can be done in C++ for efficiency